cabal-testsuite/src/Test/Cabal/Monad.hs

```diff
@@ -198,16 +206,69 @@ skipUnlessIO reason b = unless b (skipIO reason)
 skipUnless :: String -> Bool -> TestM ()
 skipUnless reason b = unless b (skip reason)
 
-expectedBroken :: Int -> TestM a
-expectedBroken t = liftIO $ do
-    putStrLn "EXPECTED FAIL"
-    E.throwIO (TestCodeKnownFail t)
-
-unexpectedSuccess :: TestM a
-unexpectedSuccess = liftIO $ do
-    putStrLn "UNEXPECTED OK"
-    E.throwIO TestCodeUnexpectedOk
+-- * Broken tests
 
+expectBroken :: Int -> TestM a -> TestM a
+expectBroken ticket m = do
+    let ticket' = IssueID ticket
+    env <- getTestEnv
+    liftIO . withAsync (runReaderT m env) $ \a -> do
+        r <- waitCatch a
+        case r of
+            Left e  -> do
+                putStrLn $ "This test is known broken, see " ++ show ticket' ++ ":"
+                print e
+                expectedBroken ticket'
+            Right _ -> do
+                unexpectedSuccess ticket'
+
+expectBrokenIf :: Bool -> Int -> TestM a -> TestM a
+expectBrokenIf True ticket m = expectBroken ticket m
+expectBrokenIf False _ m = m
+
+expectBrokenUnless :: Bool -> Int -> TestM a -> TestM a
+expectBrokenUnless b = expectBrokenIf (not b)
+
+expectedBroken :: IssueID -> IO a
```

Comment:
Can we change it to `printExpectedBroken` or `expectedBrokenMsg` or anything like this? The current name is just too close to `expectBroken` while serving a completely different purpose. I know you're not the author of these names but since you're updating it anyway...

Same for the other printing functions.

Also since you're updating this: could you replace `Int` with the type synonym for issue IDs?

Reply:
Done